File: trunk/src/cgi_python/msg_cgi/cgi_msg_add.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

import sys
import time

# 包含上级目录，即../common
sys.path.append('..')

# cgi, proto_io
from common.cgi.cgi_ret_code_def import CGI_RET_CODE
import common.cgi.cgi_my as _cgi
import common.proto_io.proto_io_tcp_client as _tcp_client

# proto
import common.proto.msg_pb2 as _msg
import common.proto.msg_sixin_pb2 as _msg_sixin
import common.proto.comm_struct_pb2 as _comm_struct
import common.proto.comm_enum_pb2 as _comm_enum

# util
from common.util.common_util import *

msgServerIP = '127.0.0.1'

def InnerProcess():
	# 1.Get HTTP params
	#'''
	strMsgContent = mycgi.getEnvValue('content', '')
	openid_md5_from = int(mycgi.getEnvValue('openid_md5_from', 0))
	openid_md5_to = int(mycgi.getEnvValue('openid_md5_to', 0))

	# for test
	'''
	strMsgContent = '你好！world'
	openid_md5_from = 123
	openid_md5_to = 456
	'''

	# 2.Params check
	if strMsgContent == '' or openid_md5_from == 0 or openid_md5_to == 0:
		logger.error('INVALID PARAM!')
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_INVALID_PARAM)		
		return True

	# 3.Build requestMsg protobuf
	stRequestMsg = _msg.Msg()

	stRequestMsgHeader = stRequestMsg.head
	stRequestMsgHeader.cmd = _msg.ADD_MSG_REQ
	stRequestMsgHeader.seq = int(time.time())

	stRequestMsgBody = stRequestMsg.msg_add_req
	msgcontent = stRequestMsgBody.msg_content
	msgcontent.content = strMsgContent.decode('utf-8')
	msgcontent.openid_md5_from = openid_md5_from
	msgcontent.openid_md5_to = openid_md5_to

	logger.debug('requestMsg: \n %s', stRequestMsg)

	# 4.Send to server, and recv responseMsg protobuf	
	ioclient = _tcp_client.ProtoIOTcpClient(msgServerIP, msgServerPort)

	dictMsg = {'requestMsg':stRequestMsg, 'responseMsg':None, 'strErrMsg':None}

	ret = ioclient.io(dictMsg)

	if ret != 0:	
		logger.error('ProtoIOTcpClient IO failed, errmsg = %s', dictMsg['strErrMsg'])
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	stResponseMsg = dictMsg['responseMsg']
	logger.debug('responseMsg: \n %s', stResponseMsg)

	if stResponseMsg.head.cmd != _msg.ADD_MSG_RES:
		logger.error('response.head.cmd=%d, unkown, fuck!!!', stResponseMsg.head.cmd)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	if stResponseMsg.head.result != _msg.E_OK:
		logger.error('response.head.result=%d, not E_OK!', stResponseMsg.head.result)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	# 5.protobuf -> json, and do reply
	msg_id = stResponseMsg.msg_add_res.msg_id
	create_ts = stResponseMsg.msg_add_res.create_ts

	mycgi.addBody('msg_id', msg_id)
	mycgi.addBody('create_ts', create_ts)

	mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_OK)
	return True





if __name__ == '__main__':
	# __file__ 获取模块所在路径
	mycgi = _cgi.myCgi(__file__, 'logging.conf')

	# logger	
	logger = mycgi.getLogger()

	InnerProcess()	

File: trunk/src/cgi/msg_cgi/cgi_msg_add.cpp
#include "cgi_msg_server.h"
#include "proto_io_tcp_client.h"
#include <time.h>
#include "msg.pb.h"

class CgiMsgAdd: public CgiMsgServer
{
public:
	CgiMsgAdd():
		CgiMsgServer(0, "config.ini", "logger.properties")
	{

	}

	bool InnerProcess()
	{
		//1.get HTTP params
		std::string strMsgContent = (std::string)GetInput().GetValue("content");
		uint64_t openid_md5_from = strtoul(((string)GetInput().GetValue("openid_md5_from")).c_str(), NULL, 10);
		uint64_t openid_md5_to = strtoul(((string)GetInput().GetValue("openid_md5_to")).c_str(), NULL, 10);

		if(strMsgContent.empty() || isAllSpace(strMsgContent.c_str()) || strMsgContent.size()>lenLimit)
		{
			DoReply(CGI_RET_CODE_INVALID_PARAM);
			LOG4CPLUS_ERROR(logger, "INVALID MSG CONTENT");
			return true;	
		}

		if(0 == openid_md5_from || 0 == openid_md5_to)
		{
			DoReply(CGI_RET_CODE_INVALID_PARAM);
			LOG4CPLUS_ERROR(logger, "INVALID OPENID MD5");
			return true;	
		}
		
		//2.Build requestMsg protobuf
		std::string strErrMsg = "";
		::hoosho::msg::Msg stRequestMsg;
		::hoosho::msg::Msg stResponseMsg;
		
		::hoosho::msg::MsgHead* stRequestMsgHead = stRequestMsg.mutable_head();
		stRequestMsgHead->set_cmd(::hoosho::msg::ADD_MSG_REQ);		
		stRequestMsgHead->set_seq(time(NULL));

		::hoosho::sixin::AddSixinReq* stMsgAddReq = stRequestMsg.mutable_msg_add_req();
		::hoosho::commstruct::MsgContent* stMsgContent = stMsgAddReq->mutable_msg_content();
		stMsgContent->set_content(strMsgContent);
		stMsgContent->set_openid_md5_from(openid_md5_from);
		stMsgContent->set_openid_md5_to(openid_md5_to);

		//3.send to server, and recv responseMsg protobuf
		common::protoio::ProtoIOTcpClient ioclient(msgServerIP, msgServerPort);

		int iRet;
		iRet = ioclient.io(stRequestMsg, stResponseMsg, strErrMsg);

		if(iRet != 0)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "ProtoIOTcpClient IO failed, errmsg = " << strErrMsg);
			return true;
		}

		LOG4CPLUS_DEBUG(logger, "response Msg: \n" << stResponseMsg.Utf8DebugString());
		//parse response
		const ::hoosho::msg::MsgHead& stHead = stResponseMsg.head();
		if(stHead.cmd() != ::hoosho::msg::ADD_MSG_RES)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.cmd="<<stHead.cmd()<<", unknown, fuck!!!");
			return true;	
		}

		if(stHead.result() != ::hoosho::msg::E_OK)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.result="<<stHead.result());
			return true;	
		}

		//4.Build strResponse Json from responseMsg protobuf
		const ::hoosho::sixin::AddSixinRes& stMsgAddRes = stResponseMsg.msg_add_res();
		GetAnyValue()["msg_id"] = int_2_str(stMsgAddRes.msg_id());
		GetAnyValue()["create_ts"] = int_2_str(stMsgAddRes.create_ts());
		
		DoReply(CGI_RET_CODE_OK);	
		
		return true;		
	}
		
};

int main()
{
	CgiMsgAdd cgi;
	if(!cgi.Run())
	{
		return -1;
	}
	return 0;
}
